lib/identity_cache/query_api.rb

```diff
@@ -132,35 +132,48 @@ def record_from_coder(coder) #:nodoc:
           klass = coder[:class].constantize
           record = klass.instantiate(coder[:attributes].dup)
 
-          coder[:associations].each {|name, value| set_embedded_association(record, name, value) } if coder.has_key?(:associations)
+          coder[:associations].each do |name, value|
+            set_embedded_association(record, name, hydrate_association_target(value))
+          end if coder.has_key?(:associations)
           coder[:association_ids].each {|name, ids| record.instance_variable_set(:"@#{record.class.cached_has_manys[name][:ids_variable_name]}", ids) } if coder.has_key?(:association_ids)
           record.readonly! if IdentityCache.fetch_read_only_records
           record
         end
       end
 
-      def set_inverse_of_cached_has_many(record, association_reflection, child_records)
-        associated_class = association_reflection.klass
-        inverse_name = record.class.cached_has_manys.fetch(association_reflection.name).fetch(:inverse_name)
+      def hydrate_association_target(dehydrated_value)
+        dehydrated_value = IdentityCache.unmap_cached_nil_for(dehydrated_value)
+        if dehydrated_value.is_a?(Array)
+          dehydrated_value.map { |coder| record_from_coder(coder) }
+        else
+          record_from_coder(dehydrated_value)
+        end
+      end
+
+      def set_inverse_of_cached_association(record, association_options, association_target)
```

Comment:
Just to make the class easier to review I'd move this method definition to after `set_embedded_association` since it is were it is used.